### PR TITLE
[SPARK-13149][SQL]Add FileStreamSource

### PR DESCRIPTION
**Pull request by zsxwing**
`FileStreamSource` is an implementation of `org.apache.spark.sql.execution.streaming.Source`. It takes advantage of the existing `HadoopFsRelationProvider` to support various file formats. It remembers files in each batch and stores it into the metadata files so as to recover them when restarting. The metadata files are stored in the file system. There will be a further PR to clean up the metadata files periodically. 

This is based on the initial work from @marmbrus.

### CODE REVIEW

**Comment by zsxwing** [comment]
@marmbrus @tdas Please take a look.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+import java.io.File
+
+import org.apache.spark.sql.StreamTest
+import org.apache.spark.sql.catalyst.util._
+import org.apache.spark.sql.execution.streaming.StreamExecution
+import org.apache.spark.sql.test.SharedSQLContext
+import org.apache.spark.util.Utils
+
+class FileStreamSourceSuite extends StreamTest with SharedSQLContext {
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Could we write these tests with the `MemorySink` instead, using `testStream`?  I'd like to be able to test things like dropped batches and restarting as well.  I would also be good to have them less coupled.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 2]
```diff
@@ -130,8 +130,9 @@ trait SchemaRelationProvider {
 trait StreamSourceProvider {
   def createSource(
       sqlContext: SQLContext,
-      parameters: Map[String, String],
-      schema: Option[StructType]): Source
+      partitionColumns: Option[StructType],
+      dataSchema: Option[StructType],
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: nit: could we order these the same as below: dataSchema, partitionColumns, parameters

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/DataStreamReaderSuite.scala [hunk 3]
```diff
@@ -33,10 +33,11 @@ object LastOptions {
 class DefaultSource extends StreamSourceProvider with StreamSinkProvider {
   override def createSource(
       sqlContext: SQLContext,
-      parameters: Map[String, String],
-      schema: Option[StructType]): Source = {
+      partitionColumns: Option[StructType],
+      dataSchema: Option[StructType],
+      parameters: Map[String, String]): Source = {
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/DataStreamReaderSuite.scala]: If we have to add stuff we should probably add a schema too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+import org.apache.spark.Logging
+import org.apache.spark.serializer.JavaSerializer
+import org.apache.spark.sql.SQLContext
+
+/**
+ * A very simple sink that stores received data on the filesystem as a text file.
+ * This is not atomic.
+ */
+class FileStreamSink(
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala]: If possible it would be nice to do this in a different PR.

File: sql/core/src/main/scala/org/apache/spark/sql/DataStreamReader.scala [hunk 5]
```diff
@@ -100,6 +100,7 @@ class DataStreamReader private[sql](sqlContext: SQLContext) extends Logging {
     val resolved = ResolvedDataSource.createSource(
       sqlContext,
       userSpecifiedSchema = userSpecifiedSchema,
+      partitionColumns = Array.empty[String],
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataStreamReader.scala]: Wait second... I think we only ever give the partition columns to a DataSource when it is coming from the hive metastore.  Given that, I don't know if we actually need to do this at all for streaming.  Could we simplify this?

**Comment by zsxwing** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    sqlContext: SQLContext,
+    metadataPath: String,
+    path: String,
+    dataSchema: Option[StructType],
+    dataFrameBuilder: Array[String] => DataFrame) extends Source with Logging {
+
+  import sqlContext.implicits._
+
+  /** Returns the schema of the data from this source */
+  override def schema: StructType = dataSchema.getOrElse(new StructType().add("value", StringType))
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: This `getOrElse` is only going to work for the text file data source, I think that for things like JSON we should probably try and initialize the source using `dataFrameBuilder` and extract the schema from there.

We should also add tests that would catch a problem here.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+  test("read from text files") {
+    val src = Utils.createTempDir("streaming.src")
+    val tmp = Utils.createTempDir("streaming.tmp")
+
+    val textSource = ResolvedDataSource.createSource(
+      sqlContext,
+      userSpecifiedSchema = None,
+      providerName = "text",
+      options = Map("path" -> src.getCanonicalPath)).asInstanceOf[FileStreamSource]
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: I'd prefer to use the `DataFrameReader/Writer` interface here so that changes to internals don't require changes to the tests.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    val textSource = ResolvedDataSource.createSource(
+      sqlContext,
+      userSpecifiedSchema = None,
+      providerName = "json",
+      options = Map("path" -> src.getCanonicalPath)).asInstanceOf[FileStreamSource]
+    val df = textSource.toDF().filter($"value" contains "keep")
+    val filtered = df
+
+    testStream(filtered)(
+      AddTextFileData(textSource, "{'c': 'drop1'}\n{'c': 'keep2'}\n{'c': 'keep3'}", src, tmp),
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: The column name here is `c` but the query is referencing `value`.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+    override def addData(): Offset = {
+      val file = Utils.tempFileWith(new File(tmp, "parquet"))
+      content.toDS().toDF().write.parquet(file.getCanonicalPath)
+      file.renameTo(new File(src, file.getName))
+      source.currentOffset + 1
+    }
+  }
+
+  test("read from text files") {
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: We should add tests that start/stop the stream as well to test recovery.  We should also consider creating a stress test.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala (hunk 7 above)
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: The problem is `DataFrameReader` doesn't return `Source`. It just returns `DateFrame`.

Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: I think that we need to have at least some tests that go through the user facing API.  I would even consider just always extracting the source from the `DataFrame` since that path is likely to be more stable.

```scala
df.queryExecution.analyzed.collect { case s: Source => s }.head
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala (hunk 6 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Updated the logic here. Now if there are any existing files, it will use them to infer the schema. And also added a test for it.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala (hunk 7 above)
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Updated the tests to use DataFrameReader

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala (hunk 8 above)
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Fixed

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala (hunk 9 above)
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Updated tests to add `StartStream` and `StopStream`. Also added a stress test.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    dataFrameBuilder: Array[String] => DataFrame) extends Source with Logging {
+
+  import sqlContext.implicits._
+
+  /** Returns the schema of the data from this source */
+  override lazy val schema: StructType = {
+    dataSchema.getOrElse {
+      val filesPresent = fetchAllFiles()
+      if (filesPresent.isEmpty) {
+        new StructType().add("value", StringType)
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Even if there are no files present, we should probably still defer to the source.  Those that can support that will work and those that don't will throw the correct error message.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+
+  import testImplicits._
+
+  case class AddTextFileData(source: FileStreamSource, content: String, src: File, tmp: File)
+    extends AddData {
+
+    override def addData(): Offset = {
+      val file = Utils.tempFileWith(new File(tmp, "text"))
+      stringToFile(file, content).renameTo(new File(src, file.getName))
+      source.currentOffset + 1
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Why +1?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala (hunk 10 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: > Those that can support that will work and those that don't will throw the correct error message.

But we need to return some `StructType` here. Any magic and defer that?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala (hunk 11 above)
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: To wait until the next batch finishes.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+      None
+    }
+  }
+
+  private def sparkContext = sqlContext.sparkContext
+
+  private val fs = FileSystem.get(sparkContext.hadoopConfiguration)
+  private val existingBatchFiles = fetchAllBatchFiles()
+  private val existingBatchIds = existingBatchFiles.map(_.getPath.getName.toInt)
+  private var maxBatchFile = if (existingBatchIds.isEmpty) -1 else existingBatchIds.max
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: nit: Isnt this `maxBatchId` rather than the file (since its an int, rather than a string)?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+  private val seenFiles = new OpenHashSet[String]
+
+  if (existingBatchFiles.nonEmpty) {
+    sqlContext.read
+        .text(existingBatchFiles.map(_.getPath.toString): _*)
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: should these be two indented rather than four indented?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+
+  private def sparkContext = sqlContext.sparkContext
+
+  private val fs = FileSystem.get(sparkContext.hadoopConfiguration)
+  private val existingBatchFiles = fetchAllBatchFiles()
+  private val existingBatchIds = existingBatchFiles.map(_.getPath.getName.toInt)
+  private var maxBatchFile = if (existingBatchIds.isEmpty) -1 else existingBatchIds.max
+  private val seenFiles = new OpenHashSet[String]
+
+  if (existingBatchFiles.nonEmpty) {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: can you add some comments on what this code is for?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala (hunk 10 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Oh I see, even `sqlContext.read.format("text").load()` doesn't work?  I would rather fix that than hardcode this here.  

For sources like parquet/json it doesn't really make sense to let them point it at an empty directory so I would rather throw an error.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala (hunk 11 above)
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: I think there is a race here then.  You've already moved the file into the directory right?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+      CheckAnswer("keep2", "keep3", "keep5", "keep6"),
+      AddParquetFileData(fileSource, Seq("drop7", "keep8", "keep9"), src, tmp),
+      CheckAnswer("keep2", "keep3", "keep5", "keep6", "keep8", "keep9")
+    )
+
+    Utils.deleteRecursively(src)
+    Utils.deleteRecursively(tmp)
+  }
+
+  test("file stress test") {
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Lets put this in its own suite since its pretty slow.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+
+import java.io.File
+
+import org.apache.spark.sql.StreamTest
+import org.apache.spark.sql.catalyst.util._
+import org.apache.spark.sql.execution.streaming.{FileStreamSource, Offset, StreamingRelation}
+import org.apache.spark.sql.test.SharedSQLContext
+import org.apache.spark.util.Utils
+
+class FileStreamSourceSuite extends StreamTest with SharedSQLContext {
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: We also need to add a test that completely recreates the `DataFrame`, which will require adding a new type of action to `StreamTest`.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala (hunk 11 above)
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Ah, right. How about adding PauseStream and ResumeStream actions? So that we can make sure no batch will run when adding data here.

**Comment by zsxwing** [comment]
@marmbrus After discussing with @tdas offline, we think there will be a lot of changes to FileStreamSource if we want to cache the content of metadata files (so that we don't need to read them in getNextBatch) and handle corrupted metadata files. So it's meaningless to review this one now. Let me just work on that now and review this one later.

**Comment by marmbrus** [comment]
How long is it going to take to implement caching?  I'd rather measure the performance of this implementation before we start optimizing.

**Comment by zsxwing** [comment]
> How long is it going to take to implement caching? I'd rather measure the performance of this implementation before we start optimizing.

Not a lot of work, actually. Just adding a <batch_id, files> map and updating it in getNextBatch.

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
Changes to this PR:

1. Add a HashMap batchToMetadata to store the metadata in memory so that we don't need to read metadata files in `getNextBatch`
2. Load metadata files and handle corrupted metadata files when finding metadata files in the constructor
3. Only text format doesn't require a schema when there are no existing files. For other cases, it will throw `IllegalArgumentException`.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 17]
```diff
@@ -195,6 +196,35 @@ trait HadoopFsRelationProvider {
...
+        files,
+        schema,
+        partitionColumns = None,
+        bucketSpec = None,
+        parameters)
+      DataFrame(sqlContext, LogicalRelation(relation))
+    }
+
+    val maybeSchema =
+      if (schema.isEmpty && providerName == "text") {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: My point was that I don't want to special case the logic here, but should instead defer to the source.  What happens if we add another source that is capable of producing a schema without any files present?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+    fs.listStatus(new Path(path))
+      .filterNot(_.getPath.getName.startsWith("_"))
+      .map(_.getPath.toUri.toString)
+  }
+
+  /**
+   * Write the metadata of a batch to disk. The file format is as follows:
+   *
+   * {{{
+   *   SPARK_VERSION
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: I think we just want a metadata version here, since multiple spark versions will likely share the same format.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 19]
```diff
@@ -0,0 +1,323 @@
...
+  }
+
+  case class AddParquetFileData(
+    source: FileStreamSource,
+    content: Seq[String],
+    src: File,
+    tmp: File) extends AddData {
+
+    override def addData(): Offset = {
+      source.currentOffset {
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Could we have gotten the same thing by checking the offset before moving the file?

Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: I think we need a `synchronized` here to make sure FileStreamSource is not scanning the files when moving the file.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala (hunk 18 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Added FileStreamSource.VERSION

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala (hunk 17 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Moved this local into FileStreamSource

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala (hunk 18 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Update this comment then.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+  private val START_TAG = "START"
+  private val END_TAG = "END"
+  val VERSION = "FILESTREAM1"
+
+  /**
+   * Parse a metadata file and return the content. If the metadata file is corrupted, it will return
+   * an empty `Seq`.
+   */
+  def readBatch(input: InputStream): Seq[String] = {
+    val lines = scala.io.Source.fromInputStream(input)(Codec.UTF8).getLines().toArray
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: We should validate the version too probably?

**Comment by marmbrus** [comment]
Small comments, otherwise LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala (hunk 20 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 21]
```diff
@@ -0,0 +1,319 @@
...
+
+    Utils.deleteRecursively(src)
+    Utils.deleteRecursively(tmp)
+  }
+
+  test("file stream source without schema") {
+    val src = Utils.createTempDir("streaming.src")
+
+    // Only "text" doesn't need a schema
+    createFileStreamSource("text", src.getCanonicalPath)
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Can we also make sure we throw a better error for this case?

```scala
scala> sqlContext.read.format("text").stream()
java.util.NoSuchElementException: key not found: path
  at scala.collection.MapLike$class.default(MapLike.scala:228)
  at scala.collection.AbstractMap.default(Map.scala:59)
  at scala.collection.MapLike$class.apply(MapLike.scala:141)
  at scala.collection.AbstractMap.apply(Map.scala:59)
  at org.apache.spark.sql.sources.HadoopFsRelationProvider$class.createSource(interfaces.scala:206)
  at org.apache.spark.sql.execution.datasources.text.DefaultSource.createSource(DefaultSource.scala:42)
  at org.apache.spark.sql.execution.datasources.ResolvedDataSource$.createSource(ResolvedDataSource.scala:107)
  at org.apache.spark.sql.DataFrameReader.stream(DataFrameReader.scala:167)
  ... 40 elided
```

Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: I would like to fix it in a separate PR since `load` throws the same error:

```
scala> sqlContext.read.format("text").load()
java.util.NoSuchElementException: key not found: path
  at scala.collection.MapLike$class.default(MapLike.scala:228)
  at org.apache.spark.sql.execution.datasources.CaseInsensitiveMap.default(ddl.scala:159)
  at scala.collection.MapLike$class.apply(MapLike.scala:141)
  at org.apache.spark.sql.execution.datasources.CaseInsensitiveMap.apply(ddl.scala:159)
  at org.apache.spark.sql.execution.datasources.ResolvedDataSource$$anonfun$10.apply(ResolvedDataSource.scala:200)
  at org.apache.spark.sql.execution.datasources.ResolvedDataSource$$anonfun$10.apply(ResolvedDataSource.scala:200)
  at scala.Option.getOrElse(Option.scala:121)
  at org.apache.spark.sql.execution.datasources.ResolvedDataSource$.apply(ResolvedDataSource.scala:200)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:129)
  ... 49 elided
```

Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Fixed the path error for `stream`

**Comment by zsxwing** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+
+  private val fs = FileSystem.get(sqlContext.sparkContext.hadoopConfiguration)
+  private var maxBatchId = -1
+  private val seenFiles = new OpenHashSet[String]
+
+  /** Cache files for each batch. The content of this map is also stored in the disk.  */
+  private val batchToMetadata = new HashMap[Long, Seq[String]]
+
+  {
+    // Restore statues from the metadata files
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: nit: statuses, not statues. :)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 23]
```diff
@@ -0,0 +1,233 @@
...
+    path: String,
+    dataSchema: Option[StructType],
+    providerName: String,
+    dataFrameBuilder: Array[String] => DataFrame) extends Source with Logging {
+
+  private val fs = FileSystem.get(sqlContext.sparkContext.hadoopConfiguration)
+  private var maxBatchId = -1
+  private val seenFiles = new OpenHashSet[String]
+
+  /** Cache files for each batch. The content of this map is also stored in the disk.  */
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: nit: Just say, Map of batch id to files.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 24]
```diff
@@ -0,0 +1,233 @@
...
+    val existingBatchFiles = fetchAllBatchFiles()
+    if (existingBatchFiles.nonEmpty) {
+      val existingBatchIds = existingBatchFiles.map(_.getPath.getName.toInt)
+      maxBatchId = existingBatchIds.max
+      // Recover "batchToMetadata" and "seenFiles" from existing metadata files.
+      existingBatchIds.sorted.foreach { batchId =>
+        val files = readBatch(batchId)
+        if (files.isEmpty) {
+          // Assert that the corrupted file must be the latest metadata file.
+          require(batchId == maxBatchId, "Invalid metadata files")
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: this should be a illegalstateexception, not illegalargumentexception since this is not based on some argument that user did.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 25]
```diff
@@ -0,0 +1,233 @@
...
+      val filesPresent = fetchAllFiles()
+      if (filesPresent.isEmpty) {
+        if (providerName == "text") {
+          // Add a default schema for "text"
+          new StructType().add("value", StringType)
+        } else {
+          throw new IllegalArgumentException("No schema specified")
+        }
+      } else {
+        // There are some existing files. Use them to infer the schema
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: nit: "." at the end.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 26]
```diff
@@ -0,0 +1,233 @@
...
+      input.close()
+    }
+  }
+}
+
+object FileStreamSource {
+
+  private val START_TAG = "START"
+  private val END_TAG = "END"
+  val VERSION = "FILESTREAM1"
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: FILESTREAM_V1

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 27]
```diff
@@ -0,0 +1,233 @@
...
+   * {{{
+   *   <FileStreamSource.VERSION>
+   *   START
+   *   -/a/b/c
+   *   -/d/e/f
+   *   ...
+   *   END
+   * }}}
+   *
+   * Note: <FileStreamSource.VERSION> means the value of `FileStreamSource.VERSION`.  Every file
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: extra space after .

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 28]
```diff
@@ -0,0 +1,233 @@
...
+    if (lines.head != VERSION) {
+      return Nil
+    }
+    if (lines(1) != START_TAG) {
+      return Nil
+    }
+    if (lines.last != END_TAG) {
+      return Nil
+    }
+    lines.slice(2, lines.length - 1).map(_.drop(1)) // Drop character "-"
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: stripPrefix("-")

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 29]
```diff
@@ -0,0 +1,233 @@
...
+   */
+  private def writeBatch(id: Int, files: Seq[String]): Unit = {
+    assert(files.nonEmpty, "create a new batch without any file")
+    val output = fs.create(new Path(metadataPath + "/" + id), true)
+    val writer = new PrintWriter(new OutputStreamWriter(output, UTF_8))
+    try {
+      // scalastyle:off println
+      writer.println(FileStreamSource.VERSION)
+      writer.println(FileStreamSource.START_TAG)
+      files.foreach(file => writer.println("-" + file))
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: would be good to make "-" a field like START_TAG. Say, PATH_PREFIX?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 30]
```diff
@@ -0,0 +1,319 @@
...
+    override def addData(): Offset = {
+      source.currentOffset {
+        val file = Utils.tempFileWith(new File(tmp, "text"))
+        stringToFile(file, content).renameTo(new File(src, file.getName))
+      } + 1
+    }
+  }
+
+  case class AddParquetFileData(
+    source: FileStreamSource,
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: incorrect indent.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 31]
```diff
@@ -0,0 +1,233 @@
...
+    }
+
+    new LongOffset(maxBatchId)
+  }
+
+  /**
+   * For test only. Run `action` and return the current offset. When `action` is running, the method
+   * guarantee that the current offset won't be changed and no new batch will be emitted.
+   */
+  def currentOffset(action: => Unit = {}): LongOffset = synchronized {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: this is not a good name. `currentOffset` cannot take a body of code. Rather better to name this something like `withBatchingLocked`
```
fileStreamSource.withBatchingLocked {
    // do something
}
```

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 32]
```diff
@@ -0,0 +1,319 @@
...
+
+  import testImplicits._
+
+  test("read from text files") {
+    val src = Utils.createTempDir("streaming.src")
+    val tmp = Utils.createTempDir("streaming.tmp")
+
+    val textSource = createFileStreamSource("text", src.getCanonicalPath)
+    val df = textSource.toDF().filter($"value" contains "keep")
+    val filtered = df
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: while this extra line?? 
why not `val filtered = textSource.toDF().filter($"value" contains "keep")`

**Comment by tdas** [comment]
Offline discussion: 
It will be good to have a unit test that documents/tests the behavior of what file source gets generated when format and/or schema is provided/not provided and existing files present / not present, what is the schema generated in the source. So that we have clear understanding of what the behavior is.

**Comment by tdas** [comment]
LGTM. New tests are great, and they make sense. Merging this!
Thanks @zsxwing and @marmbrus
